src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/CompletionItemExtensions.cs

```diff
@@ -60,15 +65,33 @@ public static VSCompletionItem ToVSCompletionItem(this CompletionItem completion
                 Detail = completion.Detail,
                 Documentation = completion.Documentation,
                 FilterText = completion.FilterText,
-                InsertText = completion.FilterText,
+                InsertText = completion.InsertText,
                 InsertTextFormat = completion.InsertTextFormat,
                 Kind = completion.Kind,
                 Label = completion.Label,
                 Preselect = completion.Preselect,
                 SortText = completion.SortText,
                 Tags = completion.Tags,
                 TextEdit = completion.TextEdit,
+                VsCommitCharacters = vsCommitCharacters,
             };
+
+            static IEnumerable<VSCommitCharacter> GetVSCommitCharacters(CompletionItem completion)
+            {
+                if (completion.CommitCharacters is null)
+                {
+                    yield break;
+                }
+
+                foreach (var commitCharacter in completion.CommitCharacters)
+                {
+                    yield return new VSCommitCharacter
+                    {
+                        Character = commitCharacter,
+                        Insert = completion.InsertTextFormat != InsertTextFormat.Snippet,
```

Comment:
Ah so we mute the characters for snippet based attributes every time, interesting.

One scenario that just popped into my head is when folks want to do `@bind:` they'll commit the completion item with `:` and have the `:` inserted. But that being said I imagine they wouldn't want the snippet experience at all in that scenario. Might be worth trying out that scenario to ensure things still work as expected.

Reply:
`@bind` is unaffected as I currently have things because it's a DirectiveAttribute instead of a `TagHelperAttribute`.

That of course raises the question of behavior for things like `@onclick`. It certainly would be a reasonable expectation for users to want `@onclick="|"`, but  it's complicated by things like `@bind:event`, `@bind-`, `@...`. Maybe leave DirectiveAttribute alone for now and file a follow-up?

Reply:
> Maybe leave DirectiveAttribute alone for now and file a follow-up?

Sounds good!  Ya I think a follow up would be, if there's no attribute parameter descriptors for the item then we can generate a snippet, otherwise if there is then we'd not. That'd probably require baking snippets into our completion provider tech though

Reply:
Filed https://github.com/dotnet/razor-tooling/issues/6125.